Module named org.objectweb.asm.tree.analysis was already on the JVMs module path

<!-- Thank you for reporting! -->

<!-- For support with crashes and issues, please make a post in #user-support channel in NeoForge Discord instead: -->
<!-- https://discord.gg/UvedJ9m -->

**Minecraft Version:** 1.21.1

**NeoForge Version:** 21.1.51

**Logs:** 

This issues has also been reported on discord: https://discord.com/channels/313125603924639766/1249305774987939900/1284093357186088970

```
2024-09-13T16:57:07.708+0200 [ERROR] [system.err] Exception in thread "main" java.lang.IllegalStateException: Module named org.objectweb.asm.tree.analysis was already on the JVMs module path loaded from ...gradle\caches\modules-2\files-2.1\org.ow2.asm\asm-analysis\9.7\e4a258b7eb96107106c0599f0061cfc1832fe07a\asm-analysis-9.7.jar but class-path contains it at location ....gradle\caches\modules-2\files-2.1\org.ow2.asm\asm-analysis\9.5\490bacc77de7cbc0be1a30bb3471072d705be4a4\asm-analysis-9.5.jar
2024-09-13T16:57:07.708+0200 [ERROR] [system.err] 	at cpw.mods.bootstraplauncher@2.0.2/cpw.mods.bootstraplauncher.BootstrapLauncher.run(BootstrapLauncher.java:139)
2024-09-13T16:57:07.709+0200 [ERROR] [system.err] 	at cpw.mods.bootstraplauncher@2.0.2/cpw.mods.bootstraplauncher.BootstrapLauncher.main(BootstrapLauncher.java:69)
```

**Steps to Reproduce:**
 1. Use the latest version 21.1.51
 2. runClient / runServer

**Description of issue:**

`cpw.mods.bootstraplauncher.BootstrapLauncher` throws an exception that as module named org.objectweb.asm.tree.analysis (9.7) was already on the JVMs module path 

`asm-analysis 9.5` is introduced by `net.neoforged:mergetool:2.0.3` e.g. in the step `neoGradleInternalUnhandledNeoFormMinecraftDependenciesForneoFormJoined1.21.1-20240808.144430` while all other parts depend on version 9.7

Even the latest versions of mergetool https://maven.neoforged.net/releases/net/neoforged/mergetool seem to depend on 9.5

There is an "ignoreList" system property used in BootstrapLauncher , but I dont know how to modify it.

--- a/common/src/main/java/net/neoforged/gradle/common/util/run/RunsUtil.java
+++ b/common/src/main/java/net/neoforged/gradle/common/util/run/RunsUtil.java
@@ -50,9 +50,7 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
 import java.util.Set;
-import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BiFunction;
-import java.util.function.BinaryOperator;
 import java.util.function.Function;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -63,8 +61,8 @@ public class RunsUtil {
         throw new IllegalStateException("Tried to create utility class!");
     }
 
-    public static String createTaskName(final String prefix, final Run run) {
-        return createTaskName(prefix, run.getName());
+    public static String createNameFor(final String prefix, final Run run) {
+        return createNameFor(prefix, run.getName());
     }
 
     public static void configure(Project project, Run run, boolean isInternal) {
@@ -91,7 +89,7 @@ public class RunsUtil {
     public static void createTasks(Project project, Run run) {
         if (!run.getIsJUnit().get()) {
             //Create run exec tasks for all non-unit test runs
-            project.getTasks().register(createTaskName(run.getName()), JavaExec.class, runExec -> {
+            project.getTasks().register(createNameFor(run.getName()), JavaExec.class, runExec -> {
                 runExec.setDescription("Runs the " + run.getName() + " run.");
                 runExec.setGroup("NeoGradle/Runs");
 
@@ -220,7 +218,7 @@ public class RunsUtil {
                 throw new InvalidUserDataException("RenderDoc can only be enabled for client runs.");
 
             final RenderDocTools renderDocTools = project.getExtensions().getByType(Subsystems.class).getTools().getRenderDoc();
-            final TaskProvider<RenderDocDownloaderTask> setupRenderDoc = project.getTasks().register(RunsUtil.createTaskName("setupRenderDoc", run), RenderDocDownloaderTask.class, renderDoc -> {
+            final TaskProvider<RenderDocDownloaderTask> setupRenderDoc = project.getTasks().register(RunsUtil.createNameFor("setupRenderDoc", run), RenderDocDownloaderTask.class, renderDoc -> {
                 renderDoc.getRenderDocVersion().set(renderDocTools.getRenderDocVersion());
                 renderDoc.getRenderDocOutputDirectory().set(renderDocTools.getRenderDocPath().dir("download"));
                 renderDoc.getRenderDocInstallationDirectory().set(renderDocTools.getRenderDocPath().dir("installation"));
@@ -347,7 +345,7 @@ public class RunsUtil {
 
     private static void createNewTestTask(Project project, String name, Run run) {
         //Create a test task for unit tests
-        TaskProvider<Test> newTestTask = project.getTasks().register(createTaskName("test", name), Test.class);
+        TaskProvider<Test> newTestTask = project.getTasks().register(createNameFor("test", name), Test.class);
         configureTestTask(project, newTestTask, run);
         project.getTasks().named("check", check -> check.dependsOn(newTestTask));
     }
@@ -663,11 +661,11 @@ public class RunsUtil {
                 .collect(Collectors.joining(File.pathSeparator)));
     }
 
-    private static String createTaskName(final String runName) {
-        return createTaskName("run", runName);
+    private static String createNameFor(final String runName) {
+        return createNameFor("run", runName);
     }
 
-    private static String createTaskName(final String prefix, final String runName) {
+    private static String createNameFor(final String prefix, final String runName) {
         final String conventionTaskName = runName.replaceAll("[^a-zA-Z0-9\\-_]", "");
         if (conventionTaskName.startsWith("run")) {
             return conventionTaskName;

--- a/userdev/src/main/java/net/neoforged/gradle/userdev/runtime/definition/UserDevRuntimeDefinition.java
+++ b/userdev/src/main/java/net/neoforged/gradle/userdev/runtime/definition/UserDevRuntimeDefinition.java
@@ -8,7 +8,6 @@ import net.neoforged.gradle.common.runtime.tasks.DownloadAssets;
 import net.neoforged.gradle.common.runtime.tasks.ExtractNatives;
 import net.neoforged.gradle.common.util.ConfigurationUtils;
 import net.neoforged.gradle.common.util.run.RunsUtil;
-import net.neoforged.gradle.dsl.common.runs.run.DependencyHandler;
 import net.neoforged.gradle.dsl.common.runtime.definition.Definition;
 import net.neoforged.gradle.dsl.common.tasks.WithOutput;
 import net.neoforged.gradle.dsl.userdev.configurations.UserdevProfile;
@@ -127,14 +126,26 @@ public final class UserDevRuntimeDefinition extends CommonRuntimeDefinition<User
         }
 
         final TaskProvider<ClasspathSerializer> minecraftClasspathSerializer = getSpecification().getProject().getTasks().register(
-                RunsUtil.createTaskName("writeMinecraftClasspath", run),
+                RunsUtil.createNameFor("writeMinecraftClasspath", run),
                 ClasspathSerializer.class,
                 task -> {
-                    this.additionalUserDevDependencies.getExtendsFrom().forEach(task.getInputFiles()::from);
-                    task.getInputFiles().from(this.additionalUserDevDependencies);
-                    task.getInputFiles().from(neoformRuntimeDefinition.getMinecraftDependenciesConfiguration());
+                    final Configuration lcpConfiguration = ConfigurationUtils.temporaryConfiguration(getSpecification().getProject(),
+                            RunsUtil.createNameFor("lcp", run));
+
+                    lcpConfiguration.extendsFrom(neoformRuntimeDefinition.getMinecraftDependenciesConfiguration());
+                    lcpConfiguration.extendsFrom(this.additionalUserDevDependencies);
+                    lcpConfiguration.extendsFrom(run.getDependencies().getRuntimeConfiguration());
+
+                    //We depend on the configuration, this ensures that if we have dependencies with different versions in the
+                    //dependency tree they are resolved to one version and are not added with different versions to the ConfigurableFileCollection
+                    //on the input files.
+                    //Additionally, we need to directly depend on the output file of the userdevClasspathElementProducer, we can not convert
+                    //this to a dependency because it would be a transform of a configurable file collection that holds the output of the task
+                    //which requires running that task, which would be a cyclic dependency.
+                    //This is a workaround to ensure that the path to the userdev classpath element is resolved correctly.
+                    //And added to the LCP, as we are now not transforming the CFC created from the output (we are not even creating a CFC)
+                    task.getInputFiles().from(lcpConfiguration);
                     task.getInputFiles().from(this.userdevClasspathElementProducer.flatMap(WithOutput::getOutput));
-                    task.getInputFiles().from(run.getDependencies().getRuntimeConfiguration());
                 }
         );
         configureAssociatedTask(minecraftClasspathSerializer);

